crates/ra_proc_macro/src/process.rs

```diff
@@ -0,0 +1,209 @@
+        let thread = ProcMacroProcessThread { handle, sender: SenderGuard(task_tx) };
+
+        Ok((thread, srv))
+    }
+
+    pub fn find_proc_macros(
+        &self,
+        dylib_path: &Path,
+    ) -> Result<Vec<(String, ProcMacroKind)>, ra_tt::ExpansionError> {
+        let task = ListMacrosTask { lib: dylib_path.to_path_buf() };
+
+        let result: ListMacrosResult = self.send_task(Request::ListMacro(task))?;
+        Ok(result.macros)
+    }
+
+    pub fn custom_derive(
+        &self,
+        dylib_path: &Path,
+        subtree: &Subtree,
+        derive_name: &str,
+    ) -> Result<Subtree, ra_tt::ExpansionError> {
+        let task = ExpansionTask {
+            macro_body: subtree.clone(),
+            macro_name: derive_name.to_string(),
+            attributes: None,
+            lib: dylib_path.to_path_buf(),
+        };
+
+        let result: ExpansionResult = self.send_task(Request::ExpansionMacro(task))?;
+        Ok(result.expansion)
+    }
+
+    pub fn send_task<R>(&self, req: Request) -> Result<R, ra_tt::ExpansionError>
+    where
+        R: TryFrom<Response, Error = &'static str>,
+    {
+        let sender = match &self.inner {
+            None => return Err(ra_tt::ExpansionError::Unknown("No sender is found.".to_string())),
+            Some(it) => it,
+        };
+
+        let (result_tx, result_rx) = bounded(0);
+
+        sender.send(Task::Request { req: req.into(), result_tx }).map_err(|err| {
+            ra_tt::ExpansionError::Unknown(format!(
+                "Fail to send task in channel, reason : {:#?} ",
+                err
+            ))
+        })?;
+
+        let res = result_rx.recv().unwrap();
+        match res {
+            Response::Error(err) => {
+                return Err(ra_tt::ExpansionError::ExpansionError(err.message));
+            }
+            _ => Ok(res.try_into().map_err(|err| {
+                ra_tt::ExpansionError::Unknown(format!(
+                    "Fail to get response, reason : {:#?} ",
+                    err
+                ))
+            })?),
+        }
+    }
+}
+
+fn client_loop(task_rx: Receiver<Task>, mut process: Process) {
+    let (mut stdin, mut stdout) = match process.stdio() {
+        None => return,
+        Some(it) => it,
+    };
+
+    loop {
+        let task = match task_rx.recv() {
+            Ok(task) => task,
+            Err(_) => break,
+        };
+
+        let (req, result_tx) = match task {
+            Task::Request { req, result_tx } => (req, result_tx),
+            Task::Close => break,
+        };
+
+        let res = match send_request(&mut stdin, &mut stdout, req) {
+            Ok(res) => res,
+            Err(_err) => {
+                let res = Response::Error(ResponseError {
+                    code: ErrorCode::ServerErrorEnd,
+                    message: "Server closed".into(),
+                });
+                if result_tx.send(res.into()).is_err() {
```

Comment:
Similarly, lets make this an unwrap, and make sure that the opposite side of the channel is not dropped to early

Reply:
Fixed. 

The recv side of this channel would not be dropped as it was passed by the task channel: 

https://github.com/rust-analyzer/rust-analyzer/blob/eb84838e2c0344acd6b007365ef7fea9fc30759e/crates/ra_proc_macro/src/process.rs#L136-L140